### PR TITLE
Include store path exact spec in the docs

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

This is niche, but deserves to be in the manual because it is describing behavior visible to the outside world, not mere implementation details.

# Context

~~I think this isn't quite ready until some other things are documented first, like the NAR format, and how file system objects can be content addressed more generally.~~

Nah let's just get this over with. It's fine.

~~I also want this to have some sort of "advanced" disclaimer.~~

Done

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by fricklerhandwerk** [comment]
Hm, wouldn't it make more sense to make the doxygen output more prominent for the purpose of reimplementing things?

**Comment by Ericson2314** [comment]
Perhaps? Right many separate functions create different types of store paths, whereas this comment wasn't associated with any one of function but described all of them / the big picture.

**Comment by Ericson2314** [comment]
I do think this should go in the main manual (though appendix/protocols section rather than this is OK) because at the end of the data it is describing Nix's behavior --- what do valid store paths look like? --- not the private details of a *specific* implementation.

File: src/libstore/store-api.cc [hunk 1]
```diff
@@ -65,85 +65,11 @@ StorePath Store::followLinksToStorePath(std::string_view path) const
...
-   It would have been nicer to handle fixed-output derivations under
-   "source", e.g. have something like "source:<rec><algo>", but we're
-   stuck with this for now...
-
-   The main reason for this way of computing names is to prevent name
-   collisions (for security).  For instance, it shouldn't be feasible
-   to come up with a derivation whose output path collides with the
-   path for a copied source.  The former would have a <s> starting with
-   "output:out:", while the latter would have a <s> starting with
-   "source:".
```
Comment (Ericson2314) [inline src/libstore/store-api.cc:146]: Note that the first paragraph of this is wrong, and the second two are misleading. That is why the new version (the historical note) looks quite different.

File: doc/manual/src/glossary.md [hunk 2]
```diff
@@ -46,9 +46,31 @@
...
+- [output-addressed store object]{#gloss-output-addressed-store-object}
+
+  A [store object] whose [store path] is determined by its contents.
+  This includes derivations, the outputs of [content-addressed derivations](#gloss-content-addressed-derivation), and the outputs of [fixed-output derivations](#gloss-fixed-output-derivation).
+
+- [(floating) content-addressing derivation]{#gloss-content-addressed-derivation}
+  A derivation whose outputs are will content addressed.
+  The "floating" adjective it often used to distinguish ones where the output content-address is not known in advanced (like reigular input-0afrom fixed output derivations,
+  Technically, this
+  which has the
```
Comment (fricklerhandwerk) [inline doc/manual/src/glossary.md:73]: This can be a separate PR. I find the current change quite messy and hard to follow, so cleaning that up is worth an entire review round on its own.

File: src/libstore/store-api.cc [hunk 3]
```diff
@@ -65,85 +65,11 @@ StorePath Store::followLinksToStorePath(std::string_view path) const
...
+/*
+The algorithm here is documented in the Store section of the Nix
```
Comment (fricklerhandwerk) [inline src/libstore/store-api.cc:69]: If we really move that to the manual, make that a concrete reference. A mechanism like [tagref](https://github.com/stepchowfun/tagref) would come in handy here... in fact, maybe we can just place an ad hoc tag reference here, even if there's no automatic processing. It can still be searched for in the source.

**Review by fricklerhandwerk — COMMENTED** [review]
I'm not convinced this is an improvement, since it breaks locality of implementation comments. Assuming the main intention of this change is to increase discoverability, an alternative would be exposing the C++ API docs in a subdirectory of the manual and linking to that piece from the regular store path page of the manual.

File: doc/manual/src/glossary.md (hunk 2 above)
Comment (Ericson2314) [reply doc/manual/src/glossary.md:73]: oh yeah oops I didn't mean to commit this. The current situation is totally broken, it will need more work.

**Comment by Ericson2314** [comment]
> But indeed it's high-level enough to have a place in the protocols chapter.

Done, it's moved there.

>  I'd want to make sure though that the high level overview doesn't duplicate too much of it.

I don't anything does?

> I'm not fully convinced yet that this is an improvement, since it breaks locality of implementation comments

Well I don't think this is very tied to the implement. The implementation works "forwards", i.e. "given some information a store object, compute its store path. This spec reads more "backwards", like a grammar for store paths. If one could magically invert the hashing function, it looks more like a *parser* for store paths than a pretty-printer.

File: src/libstore/store-api.cc (hunk 3 above)
Comment (Ericson2314) [reply src/libstore/store-api.cc]: I wrote `protocols/store-path.md` in a new version of the comment. (The new version also is clearer about outward-visible spec vs mere implementation details; the protocol documentation should *not* be updated for the latter as they are out of scope.)

**Comment by Ericson2314** [comment]
Does this look good to you now, @fricklerhandwerk?

File: doc/manual/src/protocols/store-path.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+Regular users do *not* need to know this information --- store paths can be treated as black boxes computed from the properties of the store objects they refer to.
+But for those interested in exactly how Nix works, e.g. if they are reimplementing it, this information can be useful.
+
+```bnf
+<realized-path> ::= <store-dir>/<digest>-<name>
+```
+where
+
+- `<digest>` = base-32 representation of the first 160 bits of a [SHA-256] hash of `<pre>`
```
Comment (roberth) [inline doc/manual/src/protocols/store-path.md:13]: So you're introducing metavariable `pre` here. Can't we call that the fingerprint?

File: doc/manual/src/protocols/store-path.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  (e.g. you won't get `/nix/store/<digest>-name1` and `/nix/store/<digest>-name2`, or `/gnu/store/<digest>-name1`, with equal hash parts).
+
+- `<name>` = the name of the store object.
+
+- `<store>` = the [store directory](@docroot@/store/store-path.md#store-directory)
+
+- `<type>` = one of:
+
+  - ```bnf
+    text:<r1>:<r2>:...<rN>
```
Comment (roberth) [inline doc/manual/src/protocols/store-path.md:29]: This is ambiguous as to how an empty sequence of references is represented.

File: doc/manual/src/protocols/store-path.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    Just like in the text case, we can have the store objects referenced by their paths.
+    Additionally, we can have an optional `:self` label to denote self reference.
+
+  - ```bnf
+    output:<id>
+    ```
+
+    For either the outputs built from derivations,
+    paths copied to the store hashed that area single file hashed directly, or the via a hash algorithm other than [SHA-256][sha-256].
+    (in that case "source" is used; it's silly, but it's done that way for compatibility).
```
Comment (roberth) [inline doc/manual/src/protocols/store-path.md:50]:
```suggestion
    (in that case "source" is used; this is only necessary for compatibility).
```

Compatibility isn't silly.

File: doc/manual/src/protocols/store-path.md (hunk 4 above)
Comment (roberth) [reply doc/manual/src/protocols/store-path.md:13]: The fingerprint and the store path syntax could be separate subsections.

File: doc/manual/src/protocols/store-path.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    `<id>` is the name of the output (usually, "out").
+    For content-addressed store objects, `<id>`, is always "out".
+
+- `<inner-digest>` = base-16 representation of a SHA-256 hash of `<inner-pre>`
```
Comment (roberth) [inline doc/manual/src/protocols/store-path.md:55]: I think it'd be good for each hash operation to "join" two sections of this document.
That way it's easier to see which grammar elements live together and which don't.

File: doc/manual/src/protocols/store-path.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+The `<type>` = `source:...` and `<type>` = `output:out` grammars technically overlap, in that both can represent data hashed by its SHA-256 NAR serialization.
+
+The original reason for this way of computing names was to prevent name collisions (for security).
+For instance, the thinking was that it shouldn't be feasible to come up with a derivation whose output path collides with the path for a copied source.
+The former would have an `<inner-pre>` starting with `output:out:`, while the latter would have an `<inner-pre>` starting with `source:`.
+
+Since `64519cfd657d024ae6e2bb74cb21ad21b886fd2a` (2008), however, it was decided that separting derivation-produced vs manually-hashed content-addressed data like this was not useful.
+Now, data this is to be SHA-256 + NAR-serialization content-addressed always uses the `source:...` construction, regardless of how it was produced (manually or by derivation).
+This allows freely switching between using [fixed-output derivations](@docroot@/glossary.md#gloss-fixed-output-derivation) for fetching, and fetching out-of-band and then manually adding.
+It also removes the ambiguity from the grammar.
```
Comment (roberth) [inline doc/manual/src/protocols/store-path.md:104]: Not sure if we should reference future possibilities, but
- #9259

File: doc/manual/src/SUMMARY.md.in [hunk 9]
```diff
@@ -106,6 +106,7 @@
 - [Architecture and Design](architecture/architecture.md)
 - [Protocols](protocols/index.md)
   - [Serving Tarball Flakes](protocols/tarball-fetcher.md)
+  - [Exact Store Path Specification](protocols/store-path.md)
```
Comment (roberth) [inline doc/manual/src/SUMMARY.md.in:109]:
```suggestion
  - [Store Path Specification](protocols/store-path.md)
```

Exact is mostly tautological and I don't think it needs emphasis.

File: doc/manual/src/protocols/store-path.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+## Historical Note
+
+The `type` = `"source:" ...` and `type` = `"output:out"` grammars technically overlap in purpose,
+in that both can represent data hashed by its SHA-256 NAR serialization.
+
+The original reason for this way of computing names was to prevent name collisions (for security).
+For instance, the thinking was that it shouldn't be feasible to come up with a derivation whose output path collides with the path for a copied source.
+The former would have an `inner-fingerprint` starting with `output:out:`, while the latter would have an `inner-fingerprint` starting with `source:`.
+
+Since `64519cfd657d024ae6e2bb74cb21ad21b886fd2a` (2008), however, it was decided that separting derivation-produced vs manually-hashed content-addressed data like this was not useful.
```
Comment (roberth) [inline doc/manual/src/protocols/store-path.md:113]:
```suggestion
Since `64519cfd657d024ae6e2bb74cb21ad21b886fd2a` (2008), however, it was decided that separating derivation-produced vs manually-hashed content-addressed data like this was not useful.
```

File: doc/manual/src/protocols/store-path.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+Now, data this is to be SHA-256 + NAR-serialization content-addressed always uses the `source:...` construction, regardless of how it was produced (manually or by derivation).
```
Comment (roberth) [inline doc/manual/src/protocols/store-path.md:114]: "Now, data this is to be SHA-256 "

Don't understand; looks like editing error.

File: doc/manual/src/protocols/store-path.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+This allows freely switching between using [fixed-output derivations](@docroot@/glossary.md#gloss-fixed-output-derivation) for fetching, and fetching out-of-band and then manually adding.
```
Comment (roberth) [inline doc/manual/src/protocols/store-path.md:115]: `toFile` and path interpolation still produce `source`.
```suggestion
This allows freely switching between using [fixed-output derivations](@docroot@/glossary.md#gloss-fixed-output-derivation) for fetching, and fetching out-of-band and then manually adding.
However, for compatibility, the Nix language still produces `source:`-based paths for path interpolations.
```

Comment (Ericson2314) [reply doc/manual/src/protocols/store-path.md:125]: Actually it's text hashing.

File: doc/manual/src/protocols/store-path.md [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+      - `hash` = base-16 representation of the path or flat hash of the contents of the path (or expected contents of the path for fixed-output derivations).
+
+      Note that `id` = `"out"`, regardless of the name part of the store path.
+      Also note that NAR + SHA-256 must not use this case, and instead must use the `type` = `"source:" ...` case.
+
+[Nix Archive (NAR)]: @docroot@/glossary.md#gloss-NAR
+[sha-256]: https://en.m.wikipedia.org/wiki/SHA-256
+
+## Historical Note
```
Comment (roberth) [inline doc/manual/src/protocols/store-path.md:114]:
```suggestion
### Historical Note
```

This is a note about the inner fingerprint, so let's make it a subsection of it.

File: src/libstore/store-api.cc [hunk 14]
```diff
@@ -65,85 +65,13 @@ StorePath Store::followLinksToStorePath(std::string_view path) const
...
-   collisions (for security).  For instance, it shouldn't be feasible
-   to come up with a derivation whose output path collides with the
-   path for a copied source.  The former would have a <s> starting with
-   "output:out:", while the latter would have a <s> starting with
-   "source:".
+/*
+The exact specification of store paths is in `protocols/store-path.md`
+in the Nix manual. These few functions implement that specification.
+
+If changes do these functions go behind mere implementation changes but
```
Comment (roberth) [inline src/libstore/store-api.cc:72]:
```suggestion
If changes to these functions go beyond mere implementation changes i.e.
```

**Review by roberth — APPROVED** [review]
